### PR TITLE
block mocking support for Altair

### PR DESCRIPTION
**Pull request by etan-status**
The existing functions for mocking blocks only support phase0 blocks.
This extends block mocking support to also support Altair blocks.

### CODE REVIEW

File: tests/mocking/mock_blocks.nim [hunk 1]
```diff
@@ -8,52 +8,50 @@
...
-  var previous_block_header = state.latest_block_header
-  if previous_block_header.state_root == ZERO_HASH:
-    previous_block_header.state_root = state.hash_tree_root()
-  result.message.parent_root = previous_block_header.hash_tree_root()
-
-  signMockBlock(state, result)
-
-proc mockBlockForNextSlot*(state: phase0.BeaconState): phase0.SignedBeaconBlock =
-  mockBlock(state, state.slot + 1)
+template mockImpl[S,B](): untyped =
```
Comment (zah) [inline tests/mocking/mock_blocks.nim]: Relying on explicit generic parameters is a bad habit. You'll eventually run into situations where they don't work well. Consider using regular type parameters instead (aka typedesc). The support for explicit generic parameters in the compiler is very rudimentary and you lose features such as overloading, default arguments and others.

**Review by zah — APPROVED** [review]

File: tests/mocking/mock_blocks.nim (hunk 1 above)
Comment (etan-status) [reply tests/mocking/mock_blocks.nim]: Thanks for the insightful comment! I have updated this to use `typedesc` instead.

**Comment by etan-status** [comment]
Investigating some unexpected computations when using this for work-in-progress tests.

**Comment by etan-status** [comment]
This is not sufficient for Altair, as there are some extra steps required.

See https://github.com/ethereum/consensus-specs/blob/e558618ec5077787c87f115ed8e0da8f19c1ddf6/tests/core/pyspec/eth2spec/test/helpers/block.py#L75-L105

**Comment by etan-status** [comment]
Adjusted the mocking code to more closely reflect the spec.

Leaving as draft for now as this conflicts with #2869 and until I verified that this now work properly w.r.t. new tests.

File: tests/mocking/mock_attestations.nim [hunk 2]
```diff
@@ -27,7 +27,9 @@ proc mockAttestationData(
   doAssert state.slot >= slot
 
   if slot == state.slot:
-    result.beacon_block_root = mockBlockForNextSlot(state).message.parent_root
+    let forkedState = (ref ForkedHashedBeaconState)(beaconStateFork: forkPhase0,
+      hbsPhase0: phase0.HashedBeaconState(root: hash_tree_root(state), data: state))[]
+    result.beacon_block_root = mockBlockForNextSlot(forkedState).phase0Block.message.parent_root
```
Comment (etan-status) [inline tests/mocking/mock_attestations.nim:32]: Attestations out of scope of this PR, so only doing the minimum here for compilation while retaining previous phase0-only functionality.

**Comment by etan-status** [comment]
This will also conflict with #2873 - leaving as draft until that one is merged.

**Comment by tersec** [comment]
> This will also conflict with #2873 - leaving as draft until that one is merged.

Well, #2873 merged.

File: tests/mocking/mock_attestations.nim (hunk 2 above)
Comment (tersec) [reply tests/mocking/mock_attestations.nim:32]: Agreed.

**Comment by tersec** [comment]
> Adjusted the mocking code to more closely reflect the spec.
> 
> Leaving as draft for now as this conflicts with #2869 and until I verified that this now work properly w.r.t. new tests.

https://github.com/status-im/nimbus-eth2/pull/2869 merged.

File: tests/mocking/mock_blocks.nim [hunk 3]
```diff
@@ -6,54 +6,70 @@
...
+                of forkAltair: BeaconBlockFork.Altair
+  withBlck(result):
+    blck.message.slot = slot
+    blck.message.proposer_index = get_beacon_proposer_index(state, cache, slot).get.uint64
+    blck.message.body.eth1_data.deposit_count = getStateField(state, eth1_deposit_index)
+    blck.message.parent_root = previous_block_header.hash_tree_root()
+
+  applyRandaoReveal(state, result)
+
+  if state.beaconStateFork >= forkAltair:
```
Comment (tersec) [inline tests/mocking/mock_blocks.nim]: If `state` is pre-Altair (or pre-Merge) fork epoch, but `slot` is post-Altair/Merge epoch, this will produce incorrect/incompete results.

This seems to be a result of not having `RuntimeConfig` or similar information available here.

That being said, it's at least an unambiguous improvement over the status quo, and doesn't render any existing use case worse. `unstable` currently supports phase 0, and that's exactly as well-supported with this PR. When the block fork matches the state fork, it also works fine.

It might be worth putting a TODO in to ensure this edge case is documented though, for when it makes sense to address the block fork != state fork case.

**Review by tersec — APPROVED** [review]

File: tests/mocking/mock_blocks.nim (hunk 3 above)
Comment (etan-status) [reply tests/mocking/mock_blocks.nim]: Thanks for catching this! The other parts of the test framework typically assume `defaultRuntimeConfig`, so the fork information is available. Will revise.

**Comment by etan-status** [comment]
This should now also work at the fork from phase0 to altair.
